[Drop-In UI] Reset SharedApp state when MapboxNavigation is destroyed

tomaszrybakiewicz opened the pull request:
Closes https://github.com/mapbox/mapbox-navigation-android/issues/6000

### Description

Changes:
- Introduced `StateResetController` that resets `SharedApp.store` state when `TripSessionState` is `STOPPED` or the `MapboxNavigation` gets destroyed.
- ~Refactored `LocationPermissionComponent` to ensure that LocationPermissions are checked every time parent Activity enters STARTED lifecycle.~ (will add this in a separate PR)
- Encapsulated trip session and replay trip session management in the `TripSessionComponent`.
- Moved `ReplayRouteTripSession` to _dropin_ module.

### Screenshots or Gifs

_(recording of the QA Test App captured on Pixel 6)_

https://user-images.githubusercontent.com/2678039/178306050-162a5079-ac72-40c4-a4e2-4e37d74d1f8c.mp4

### Known issues

CRASH when transitioning from Active Guidance state with alternative routes to Free Drive state (https://github.com/mapbox/mapbox-navigation-android/issues/5985)

> note: this crash doesn't happen when navigating a single route without alternatives

abhishek1508 commented:
Using Android 11 

When click on end navigation via Notification -> Restart the app -> Select a new destination -> Start active navigation

#### Expected
The replay engine should start simulating the route

#### Actual
The replay engine even though enabled doesn't simulate the route

https://user-images.githubusercontent.com/9770186/178791160-fbed3fc8-39d7-4f11-ae05-8203dc6c058b.mp4

/cc @tomaszrybakiewicz

abhishek1508 commented:
Looking at this implementation, I am 🤔 if we do actually need to determine `Refactored LocationPermissionComponent to ensure that LocationPermissions are checked every time parent Activity enters STARTED lifecycle.` This is good to have but I am not sure of the implications of permissions as they are usually tricky and require thorough testing. It gets even more complicated when we need additional permissions like WRITE/READ Storage. We can capture this as a separate ticket and have a proper requirement listed out before attempting this.

In case of happy path we end navigation by clicking on end navigation button in the info panel and the following code gets implemented:
```
binding.endNavigation.setOnClickListener {
    store.dispatch(RoutesAction.SetRoutes(emptyList()))
    store.dispatch(DestinationAction.SetDestination(null))
    store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
}
```

Why not do the same when end navigation on trip notification is clicked. We can observe that event from `MapboxTripNotification` and execute the same logic. This would keep it super simple without touching permissions. 

@tomaszrybakiewicz

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/location/LocationPermissionComponent.kt:
```diff
@@ -100,6 +67,21 @@ internal class LocationPermissionComponent(
+            onPermissionsResult(isGranted)
+        } else {
+            launcher?.launch(LOCATION_PERMISSIONS)
+            //  ActivityResultLauncher result is dispatched in ActivityResultCallback
```
```suggestion
```

nitpicking: this comment does not help much

abhishek1508 replied:
The comment looks fine to me. Although I don't think this is necessary and rather add the type to the `launcher`

Change

```
private val launcher = try{ }
```

to

```
private val launcher: ActivityResultLauncher = try { }
```

kmadsen replied:
suggestion is to delete it because it is stating the obvious in my opinion. kind of qualifies for redundant constructs
https://kotlinlang.org/docs/coding-conventions.html#documentation-comments

it stood out to me because of the placement. i never see comments describe the code above. if it's a good tl;dr for the [activity result launcher](https://developer.android.com/training/basics/intents/result), another suggestion is just reorder it
```
            //  ActivityResultLauncher result is dispatched in ActivityResultCallback
            launcher?.launch(LOCATION_PERMISSIONS)
```

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/tripsession/TripSessionComponent.kt:
```diff
@@ -0,0 +1,96 @@
+import kotlinx.coroutines.launch
+
+/**
+ * Component class that is responsible for starting and stopping [MapboxNavigation] trip session
+ * and replay trip session when [TripSessionStarterState] changes.
```
What's the reason for moving this out of the `TripSessionStarterStateController`?

Similar to `AudioGuidanceStateController`, the reason the logic lived in a state controller is so that it can be shared with android auto.

If it lives in a component, then android auto will need to add it to the `MainCarSession` and I think there will be race conditions because there can be 2 `TripSessionComponent` trying to control the trip session state.

kmadsen replied:
Can this be a different PR? One of the next steps with the TripSessionStarterStateController, is to use it with Android Auto. I'm concerned that this change will make that harder

abhishek1508 replied:
Agreed, I remembered we had this discussion earlier and we decided not to host `TripSessionComponent` to avoid possible race conditions with android auto

abhishek1508 replied:
> We must start or stop the trip session on NavigationView lifecycle `Lifecycle.State.STARTED`.

Why is this required? We start the trip session when Drop-In UI is instantiated and permissions are granted. `mapboxNavigation` stops trip session automatically when `end navigation` on the notification is clicked. Trying to wrap my head around as to why `We must start or stop` based on lifecycle changes?

tomaszrybakiewicz replied:
> Why is this required?

I answered this in a comment below https://github.com/mapbox/mapbox-navigation-android/pull/6039#issuecomment-1183792520

kmadsen replied:
Short answer is `TripSessionComponent` is added to host the `Lifecycle` dependency which is not available in a `TripSessionStarterStateController`. Is that right?

abhishek1508 commented:
> https://github.com/mapbox/mapbox-navigation-android/pull/6039#issuecomment-1183481098

Looking at it further, it makes sense to hook onto `TripSessionState` and reset the `store` rather than hooking it onto a click. However, still trying to understand why do we need to implement
`Refactored LocationPermissionComponent to ensure that LocationPermissions are checked every time parent Activity enters STARTED lifecycle.` in this PR

@tomaszrybakiewicz

Zayankovsky commented on libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/controller/StateResetController.kt:
```diff
@@ -0,0 +1,37 @@
+        // TripSessionState might change in quick succession.
+        // Debouncing it by 100 ms ensures we skip transient values.
+        var prevState = mapboxNavigation.getTripSessionState()
+        mapboxNavigation.flowTripSessionState().debounce(100).observe { newState ->
```
The main concern with `debounce` is what if we have a slow device and more than 100ms passes when you change the replay setting. We will reset `SharedApp` state, but will we go back to the previous state after a new trip session is started?

abhishek1508 replied:
I am still curious

```
mapboxNavigation.flowTripSessionState().debounce(100).observe { newState ->
    if (prevState == TripSessionState.STARTED && newState == TripSessionState.STOPPED) {
        // we only reset Store state when TripSessionState switches from STARTED to STOPPED.
        // mapboxNavigation.setNavigationRoutes(emptyList())
        store.reset()
    }
    prevState = newState
}
```
`store.reset()` resets the `State()` such that `State.routes` is now `RoutesState.Empty`. In this case what is then setting the route back? Probably it's because routes on mapboxNavigation is still available

tomaszrybakiewicz replied:
@abhishek1508 yup. That's exactly what happens. When a user clicks the "END NAVIGATION" action in the foreground notification, MapboxNavigation stops the trip session (and all background services). but it does not remove the assigned route, hence the need for `mapboxNavigation.setNavigationRoutes(emptyList())` call.

tomaszrybakiewicz replied:
@Zayankovsky I agree with you. I do not like the use of `debounce` here. It does feel hacky. However, it allows us to skip the transient TripSession state when toggling between: 

```
MapboxNavigation.startTripSession() -> MapboxNavigation.stopTripSession() -> MapboxNavigation.startReplayTripSession()
```

and vice versa.

Zayankovsky replied:
Resolving the thread, as I don't see how this can be implemented in a better way at the moment.

kmadsen commented on libnavui-app/src/main/java/com/mapbox/navigation/ui/app/internal/controller/StateResetController.kt:
```diff
@@ -0,0 +1,37 @@
+    override fun onAttached(mapboxNavigation: MapboxNavigation) {
+        super.onAttached(mapboxNavigation)
+
+        // TripSessionState might change in quick succession.
```
Might? or does the debounce fix an issue you found while testing?

tomaszrybakiewicz replied:
TripSessionState changes in quick succession only when the replayer is enabled/disabled. Since replayer usage is optional ---> "might".

kmadsen replied:
Can you update the comment to be specific?
```suggestion
        // TripSessionState is triggered rapidly when enabling and disabling replay when calling `MapboxNavigation`; `startTripSession, `stopTripSession`, `startReplayTripSession`
```

I'm thinking that is an issue with `startReplayTripSession`, and we may be able to remove the debounce if we make it so calling `startReplayTripSession` does not require a `MapboxNavigation.stopTripSession()` call.

`startReplayTripSession` is [experimental for this reason](https://github.com/mapbox/mapbox-navigation-android/blob/9f849f7f3051c85cfc6808474e8734ae1b482931/libnavigation-core/src/main/java/com/mapbox/navigation/core/MapboxNavigation.kt#L594-L597), we may want to update how it is used.

tomaszrybakiewicz replied:
Correct. Once we patch how `startReplayTripSession` works, we'll remove the debounce.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/location/LocationPermissionComponent.kt:
```diff
@@ -54,41 +52,10 @@ internal class LocationPermissionComponent(
-            // The view model attaches, and then launches a coroutine to collect actions.
-            // The LocationPermissionComponent surfaces this issue because it is not owned by
-            // a coordinator and flowable binder. This issue was also difficult to reproduce on
-            // all devices. Launching a coroutine to update the state is a temporary solution.
```
Although, it's not clear why it should be part of this PR.. IMO this can be an entirely separate change, it does not seem related to the "Reset SharedApp state when MapboxNavigation is destroyed"

abhishek1508 replied:
If you would see, I am allowing for location permissions `Only this time`. So, when I start the app again, why is it not asking me for location permissions? It seems like I already have the permission.

https://user-images.githubusercontent.com/9770186/178817858-ae5b059d-2cfd-45ce-9f80-b9073636a191.mp4

How do we handle the use case where I switch off my location and Drop-In UI doesn't works

https://user-images.githubusercontent.com/9770186/178817955-362ddd0f-a3b4-4e86-ac72-74574493d10b.mp4

@tomaszrybakiewicz

tomaszrybakiewicz replied:
Reverted changes to this file. Resolving this conversation.

tomaszrybakiewicz commented:
> Using Android 11
> 
> When click on end navigation via Notification -> Restart the app -> Select a new destination -> Start active navigation
> 
> #### Expected
> The replay engine should start simulating the route
> 
> #### Actual
> The replay engine even though enabled doesn't simulate the route
> 

@abhishek1508 That's a bug in the QA Test App. The `MapboxNavigationViewCustomizedActivity` only updates the `ENABLE REPLAY` switch value in onCreate. When `NavigationView` state is reseted, `MapboxNavigationViewCustomizedActivity` doesn't know about it and does not flip the switch back to `OFF` state.
I think moving the `bindSwitch(menuBinding.toggleReplay` to either onStart or onResume should do the trick.

abhishek1508 commented:
There are too many questions in the PR. Summarizing them so we can all agree to one or the other in order to get the PR closer to approval

- Are we okay with the use of debounce? @kmadsen @Zayankovsky 
- Are we okay we refactoring the `LocationPermissionComponent` in this PR?
I think we should address this in a separate PR so we can answer questions like https://github.com/mapbox/mapbox-navigation-android/pull/6039#discussion_r920440150 and not make this complicated.
- Do we need to hook start and stop of trip session with lifecycle? If this is mandatory, how do we plan to resolve implications of using `TripSessionComponent` in `Android-Auto` making sure that AA works as it should. @kmadsen Can you please work with @tomaszrybakiewicz to resolve this?

kmadsen commented:
@tomaszrybakiewicz have you tried alternative approaches to this? When reading the root cause of the issue, it seems this issue could be resolved without risking regressions with location permissions and app state.

> SharedApp is a singleton, which can outlive MapboxNavigation. So when MapboxNavigation is created again it will be in Free Drive state, but SharedApp may be in Active Navigation state, since that was the state when MapboxNavigation was destroyed. That causes views to display empty data."
^ The issue description https://github.com/mapbox/mapbox-navigation-android/issues/6000

And then the work around downstream was rather small https://github.com/mapbox/1tap-android/pull/2064
``` kotlin 
        MapboxNavigationApp.registerObserver(
            object : MapboxNavigationObserver {

                override fun onAttached(mapboxNavigation: MapboxNavigation) {
                    // Nothing to do
                }

                override fun onDetached(mapboxNavigation: MapboxNavigation) {
                    SharedApp.store.dispatch(RoutesAction.SetRoutes(emptyList()))
                    SharedApp.store.dispatch(DestinationAction.SetDestination(destination = null))
                    SharedApp.store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
                }
            },
        )
```

Why is this downstream change insufficient for the SDK? [We have discussed](https://github.com/mapbox/mapbox-navigation-android/pull/5826#discussion_r878518993) making the `SharedApp` a `MapboxNavigationObserver`. If it was one, then it could reset the the state when `onDetached` is called.

tomaszrybakiewicz commented:
> @tomaszrybakiewicz have you tried alternative approaches to this? When reading the root cause of the issue, it seems this issue could be resolved without risking regressions with location permissions and trip session state 🤔 .
> 
> > SharedApp is a singleton, which can outlive MapboxNavigation. So when MapboxNavigation is created again it will be in Free Drive state, but SharedApp may be in Active Navigation state, since that was the state when MapboxNavigation was destroyed. That causes views to display empty data."
> 
> ^ The issue description #6000
> 
> And then the work around downstream was rather small [mapbox/1tap-android#2064](https://github.com/mapbox/1tap-android/pull/2064)
> 
> ```kotlin
>         MapboxNavigationApp.registerObserver(
>             object : MapboxNavigationObserver {
> 
>                 override fun onAttached(mapboxNavigation: MapboxNavigation) {
>                     // Nothing to do
>                 }
> 
>                 override fun onDetached(mapboxNavigation: MapboxNavigation) {
>                     SharedApp.store.dispatch(RoutesAction.SetRoutes(emptyList()))
>                     SharedApp.store.dispatch(DestinationAction.SetDestination(destination = null))
>                     SharedApp.store.dispatch(NavigationStateAction.Update(NavigationState.FreeDrive))
>                 }
>             },
>         )
> ```
> 
> Why is this downstream change insufficient for the SDK? [We have discussed](https://github.com/mapbox/mapbox-navigation-android/pull/5826#discussion_r878518993) making the `SharedApp` a `MapboxNavigationObserver`. If it was one, then it could reset the the state when `onDetached` is called and be setup when `onAttached` is called.

@kmadsen There is no regressions risk with the location permissions check I introduced. Besides, I already reverted changes to both the `TripSessionStarterState` and the `LocationPermissionComponent` so we can ignore this. 

The patch for the permissions check on app foreground I will introduce in a separate PR.

There are actually 3 issues that this PR solves:
1. `SharedApp.state` reset when `MapboxNavigation` instance gets destroyed. _(this problem is described in #6000  description)_ 
   This is fixed in `StateResetController.onDetached()`. 
2. `StaredApp.state` revert to the FreeDrive state when user clicks on the foreground notification's "END NAVIGATION" action.
   This is done by the `mapboxNavigation.flowTripSessionState()` observer in `StateResetController.onAttached()`.
3. Clicks on the foreground notification's "END NAVIGATION" action, stops MapboxNavigation trip session. This session needs to be restarted when application is foregrounded. ([this problem is demonstrated in issue first comment](https://github.com/mapbox/mapbox-navigation-android/issues/6000#issuecomment-1172702451))
   This is done by the `lifecycle.repeatOnLifecycle(Lifecycle.State.STARTED)` coroutine in `TripSessionComponent.onAttached`.

kmadsen commented:
> Besides, I already reverted changes to both the TripSessionStarterState and the LocationPermissionComponent so we can ignore this.

Ok noticing the `LocationPermissionComponent` 👍 

Also 👍  to the addition of `StateResetController`

Am still 🤔  on `TripSessionComponent` because that appears to have issues.

kmadsen commented on libnavui-dropin/src/main/java/com/mapbox/navigation/dropin/component/tripsession/TripSessionComponent.kt:
```diff
@@ -0,0 +1,96 @@
+ */
+@ExperimentalPreviewMapboxNavigationAPI
+@SuppressLint("MissingPermission")
+internal class TripSessionComponent(
```
Will you test this class with android auto?

Specifically, I want to be able to use this component delete this code here
https://github.com/mapbox/mapbox-navigation-android/blob/9f849f7f3051c85cfc6808474e8734ae1b482931/qa-test-app/src/main/java/com/mapbox/navigation/qa_test_app/car/MainCarSession.kt#L128-L135

kmadsen replied:
OK adding TripSessionComponent to this epic here https://github.com/mapbox/navigation-sdks/issues/1780

kmadsen reviewed (approved):
Thanks for addressing feedback. Nice work digging and answering questions! 🚢
